jujubigdata/utils.py

```diff
@@ -630,3 +636,68 @@ def spec_matches(local_spec, remote_spec):
         if v != remote_spec.get(k):
             return False
     return True
+
+
+def get_ip_for_interface(network_interface):
+    """
+    Helper to return the ip address of this machine on a specific
+    interface.
+
+    @param str network_interface: either the name of the
+    interface, or a CIDR range, in which we expect the interface's
+    ip to fall. Also accepts 0.0.0.0 (and variants, like 0/0) as a
+    special case, which will simply return what you passed in.
+
+    """
+    if network_interface.startswith('0') or network_interface == '::':
+        # Allow users to reset the charm to listening on any
+        # interface.  Allow operators to specify this however they
+        # wish (0.0.0.0, ::, 0/0, etc.).
+        return network_interface
+
+    # Is this a CIDR range, or an interface name?
+    is_cidr = len(network_interface.split(".")) == 4 or len(
+        network_interface.split(":")) == 8
+
+    if is_cidr:
+        interfaces = netifaces.interfaces()
+        for interface in interfaces:
+            for ip_version in netifaces.AF_INET, netifaces.AF_INET6:
+                try:
+                    ip = netifaces.ifaddresses(
+                        interface)[ip_version][0]['addr']
+                except KeyError:
+                    continue
+
+                if ip.startswith('fe80'):
+                    # ipaddress doesn't consider fe80::... to be an ip
+                    # address, even though interfaces on physical
+                    # hardware will list this as their ipv6 ip. Just
+                    # skip over this issue for now.
+                    continue
```

Comment:
This should be something more like this:

```python
for interface in interfaces:
    for ip_version in (netifaces.AF_INET, netifaces.AF_INET6):
        for address in netifaces.ifaddresses(interface)[ip_version]:
            ip = ipaddress.ip_address(address['addr'].decode('utf8') if hasattr(address['addr'], 'decode') else address['addr'])
            if ip.is_link_local:
                continue
            if ip in ipaddress.ip_network(network_interface):
                return ip
```

Specifically:
* Loop over all returned addresses, don't just try the first
* Use `is_link_local`
* I like `hasattr(..., 'decode')` better than `encode` / `decode`, but that's more personal preference